### DISCUSSION TITLE
How I can get tree ?

### DISCUSSION THREAD

**Discussion by jayaemekar** [Gremlinq/ExRam.Gremlinq#934] [Q&A] [answered]
I checked the code and didn`t find implementation of Tree Step.

Let`s say we have Employee Vertex and a Manages Edge ..,
If i want to get all the tree of one Manager with TreeStep would be something like ...

what I need is vertexes with "parent Id" or use the ".tree()" step
g.V('empid','pk').hasLabel('empVertex').repeat(out('ChildMemberEdge').hasLabel('empvertex')).emit().tree()
but I don't find how to do that with gremlinQ

The return type I guess should be ValueTask<(EmployeeVertex, IEnumerable<EmployeeVertex>)[]>

**Comment by danielcweber** [comment]
From the provided required result type (`ValueTask<(EmployeeVertex, IEnumerable)[]>`) it appears you don't require a whole deserialization of a tree structure, but rather a collection of all reachable vertices from a root vertex. If that's the case, a projection is probably what you want:

```
g
  .V('empid','pk')
  .hasLabel('empVertex')
  .project('root', 'nodes')
  .by(identity())
  .by(__
    .repeat(out('ChildMemberEdge').hasLabel('empvertex'))
    .emit()
    .fold())
```

Is my assumption correct? If not, there's unfortunately no tree step support in Gremlinq.

**Reply by jayaemekar** [reply]
Thanks for response. How I can have this query using ExRam.Gremlinq.Providers.CosmosDb, Some functions from below one are not available

g
  .V('empid','pk')
  .hasLabel('empVertex')
  .project('root', 'nodes')
  .by(identity())
  .by(__
    .repeat(out('ChildMemberEdge').hasLabel('empvertex'))
    .emit()
    .fold())

**Comment by danielcweber** [comment]
What exactly is missing ?

**Reply by jayaemekar** [reply]
When I tried to convert the query in the format,

`g
.V(uniqueId)
.OfType<EmpVertex>()
.Project(p => p.ToDynamic())	
.Identity()
.Loop(__ => __.Emit().Repeat(__ => __.OutE().InV()))
.Emit()
.Fold();`

Not sure how I can use **By Identity and Project** function in this query which give me the result of root and childnodes

**Reply by anheri** [reply]
cosmos db does not support the emit() step

**Reply by anheri** [reply]
I would like to see the tree() feature in gremlinq aswell as cosmos db, do support the tree step

**Comment by danielcweber** [comment]
@jayaemekar Have a look at [e.g. this unit test](https://github.com/ExRam/ExRam.Gremlinq/blob/42050476d607eeada1f235b684b4f9a0c624ffb2/test/Core.Tests/QueryExecutionTest.cs#L2097) to get an idea of how to use Project...By... correctly.

**Comment by tomiasp** [comment]
Hi, this project() by() out() fold() pattern is 10x more expensive for same result as tree() operation would be. With cosmos, we are using this mentioned pattern, and for example an query to find managers for employees costs 375 RU's. Similar query with tree() operation costs 40 RUs...

**Reply by gregorvilkner** [reply]
@danielcweber, thank you for adding the Tree() operator. We've been using Project() extensively throughout the past, but found that for more nested traversals the Tree() operator gives us the same information at a 50x performance boost. Very well done.

**Comment by tomiasp** [comment, marked as answer]
@jayaemekar btw, you can now find the Tree() -operator in the newest release.
